ops/testing.py

```diff
@@ -187,6 +187,16 @@ class TestCharm(self._charm_cls):
         TestCharm.__name__ = self._charm_cls.__name__
         self._charm = TestCharm(self._framework)
 
+    def reinitialize_charm(self) -> None:
+        """Re-initialize the charm under test."""
+        if self._charm is None:
+            raise RuntimeError('cannot call the reinitialize_charm method before begin')
+
+        self._framework = framework.Framework(
+            self._storage, self._charm_dir, self._meta, self._model)
+        self._charm = None
+        self.begin()
```

Comment:
Are we sure we want a begin() call here?

Reply:
Umm. Better than begin_with_initial_hooks, I guess...
This is a `reinitialize` method -> we want the charm to be reinitialized. 
We do also want the possibility to intervene before the charm's `__init__` is run (to simulate a relation, some config change, etc...). But we can't simulate those things unless there is a framework, I suppose.
Do we need to decouple the framework setup and the charm's init?

Reply:
Or maybe we need not to reinit the framework, unless the user asks us to?
My approach before this feature was (is) to `forget()` each object that had a handle, then reinitialize the charm (and the charm only). The framework remained.